File: DataEngineering/07_Data_Transformation/projects/python/04_currency_converter_pipeline/main.py
import os
import requests
import pandas as pd
import logging
from typing import Dict
from tenacity import retry, wait_exponential, stop_after_attempt

logging.basicConfig(level=logging.INFO, format="%(asctime)s [%(levelname)s] %(message)s")
logger = logging.getLogger(__name__)

class CurrencyConverterPipeline:
    """
    A robust currency converter utilizing external APIs for live exchange rates.
    """
    def __init__(self, base_currency: str = "USD"):
        self.base_currency = base_currency
        self.api_key = os.getenv("EXCHANGE_RATE_API_KEY", "demo_key")
        self.api_url = f"https://v6.exchangerate-api.com/v6/{self.api_key}/latest/{self.base_currency}"

    @retry(wait=wait_exponential(multiplier=1, min=2, max=10), stop=stop_after_attempt(3))
    def fetch_exchange_rates(self) -> Dict[str, float]:
        logger.info(f"Fetching exchange rates for base {self.base_currency}")
        try:
            if self.api_key == "demo_key":
                logger.warning("Using demo key. Returning mocked rates.")
                return {"EUR": 0.9, "GBP": 0.75, "JPY": 110.0, "USD": 1.0}
            
            response = requests.get(self.api_url, timeout=10)
            response.raise_for_status()
            data = response.json()
            return data.get("conversion_rates", {})
        except requests.exceptions.RequestException as e:
            logger.error(f"Failed to fetch exchange rates: {e}")
            raise

    def process(self, df: pd.DataFrame) -> pd.DataFrame:
        logger.info("Processing currency conversion.")
        rates = self.fetch_exchange_rates()
        
        def convert_to_base(row):
            currency = row['currency']
            amount = row['amount']
            rate = rates.get(currency)
            if rate is None or rate == 0:
                logger.warning(f"Unknown rate for currency {currency}. Returning original amount.")
                return amount
            return amount / rate

        df[f'{self.base_currency.lower()}_amount'] = df.apply(convert_to_base, axis=1)
        return df

if __name__ == "__main__":
    pipeline = CurrencyConverterPipeline(base_currency="USD")
    
    raw_data = pd.DataFrame({
        'tx_id': [1, 2, 3], 
        'amount': [100.0, 50.0, 200.0], 
        'currency': ['EUR', 'GBP', 'USD']
    })
    
    try:
        processed_df = pipeline.process(raw_data)
        logger.info(f"\nConverted Data:\n{processed_df}")
    except Exception as e:
        logger.error(f"Pipeline execution failed: {e}")

File: DataEngineering/07_Data_Transformation/projects/golang/04_currency_converter_pipeline/main.go
package main

import (
	"context"
	"encoding/json"
	"fmt"
	"log"
	"net/http"
	"os"
	"time"
)

type Transaction struct {
	ID       int     `json:"id"`
	Amount   float64 `json:"amount"`
	Currency string  `json:"currency"`
}

type ConvertedTransaction struct {
	Transaction
	USDAmount float64 `json:"usd_amount"`
}

type ExchangeRateClient struct {
	APIKey string
	Client *http.Client
}

func NewExchangeRateClient(apiKey string) *ExchangeRateClient {
	return &ExchangeRateClient{
		APIKey: apiKey,
		Client: &http.Client{Timeout: 10 * time.Second},
	}
}

func (c *ExchangeRateClient) FetchRates(ctx context.Context, baseCurrency string) (map[string]float64, error) {
	if c.APIKey == "demo_key" || c.APIKey == "" {
		log.Println("Using mock exchange rates due to missing/demo API key")
		return map[string]float64{"EUR": 0.9, "GBP": 0.75, "USD": 1.0}, nil
	}

	url := fmt.Sprintf("https://v6.exchangerate-api.com/v6/%s/latest/%s", c.APIKey, baseCurrency)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}

	resp, err := c.Client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("unexpected status code: %d", resp.StatusCode)
	}

	var result struct {
		ConversionRates map[string]float64 `json:"conversion_rates"`
	}
	if err := json.NewDecoder(resp.Body).Decode(&result); err != nil {
		return nil, err
	}

	return result.ConversionRates, nil
}

func main() {
	apiKey := os.Getenv("EXCHANGE_RATE_API_KEY")
	client := NewExchangeRateClient(apiKey)

	rates, err := client.FetchRates(context.Background(), "USD")
	if err != nil {
		log.Fatalf("Failed to fetch rates: %v", err)
	}

	txs := []Transaction{
		{ID: 1, Amount: 100.0, Currency: "EUR"},
		{ID: 2, Amount: 50.0, Currency: "GBP"},
		{ID: 3, Amount: 150.0, Currency: "USD"},
	}

	var processed []ConvertedTransaction
	for _, tx := range txs {
		rate, ok := rates[tx.Currency]
		if !ok || rate == 0 {
			log.Printf("No rate found for currency %s, skipping", tx.Currency)
			continue
		}

		usdAmount := tx.Amount / rate

		processed = append(processed, ConvertedTransaction{
			Transaction: tx,
			USDAmount:   usdAmount,
		})
	}

	out, _ := json.MarshalIndent(processed, "", "  ")
	fmt.Println("Processed Transactions:")
	fmt.Println(string(out))
}
